grammar.js

```diff
@@ -144,28 +144,58 @@ module.exports = grammar({
     false: $ => 'false',
 
     character_literal: $ => token(seq(
-      "'",
+      '\'',
       repeat1(choice(
         /[^\\'\n]/,
         /\\./,
         /\\\n/
       )),
-      "'"
+      '\''
     )),
 
-    string_literal: $ => token(choice(
-      seq('"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)), '"'),
-      // TODO: support multiline string literals by debugging the following:
-      // seq('"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)), '"', '+', /\n/, '"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)))
-    )),
-
-    text_block: $ => token(
-      seq(
-        /"""\s*\n/,
-        optional(/([^"]|("[^"])|(""[^"]))+/),
-        '"""'
-      )
+    string_literal: $ => choice($._string_literal, $._multiline_string_literal),
+    _string_literal: $ => seq(
+      '"',
+      repeat(choice(
+        $.string_fragment,
+        $.escape_sequence,
+      )),
+      '"'
     ),
+    _multiline_string_literal: $ => seq(
+      '"""',
+      repeat(choice(
+        alias($._multiline_string_fragment, $.multiline_string_fragment),
+        $._escape_sequence,
+      )),
+      '"""'
+    ),
+    // Workaround to https://github.com/tree-sitter/tree-sitter/issues/1156
+    // We give names to the token() constructs containing a regexp
+    // so as to obtain a node in the CST.
+    //
+    string_fragment: $ =>
+      token.immediate(prec(1, /[^"\\]+/)),
+    _multiline_string_fragment: () =>
```

Comment:
@amaanq Is there a specific reason why `_multiline_string_fragment` is not of the form `token(...)`? We had ` text_block: $ => token(...)` before.

The consequence is that the `_multiline_string_fragement` node is now a product type instead of a leaf type.

Compare 
```
  {
    "type": "multiline_string_fragment",
    "named": true,
    "fields": {}
  }
```
```
 {
   "type": "text_block",
   "named": true
 }
```

I'm writing an "AST generator" generator which generates the CST into an AST based on those node types, and I'm acting differently based on which type of node it is, that is, sum type, product type or leaf type. And the behavior of the grammar changed compared to the previous version.

Reply:
Just tried adding the `token` before, but then a test fails. With `token.immediate` it works though. I'm not familiar enough yet with the grammar rules if that would be okay.

Reply:
I'm not gonna lie, I forgot why. I use this style in several grammars of mine and I derived inspiration from another grammar.. `token.immediate` can have weird side effects causing annoying edge case bugs so be careful with that one.

Reply:
Try token(prec(-1 or prec(1 and see how those vary